File: packages/nativescriptui/src/app/core/services/backend.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';
import { catchError, finalize } from 'rxjs/operators';
import { Observable, throwError } from 'rxjs';
import { isAndroid, isIOS } from "@nativescript/core";

@Injectable({
  providedIn: "root"
})

export class BackendService {

  token: string = null;

  constructor(private http: HttpClient) 
  { 

  }
  
  public setToken(token: string)
  {
    this.token = token;
  }

  public clearToken()
  {
    this.token = null;    
  }

  public getDataFromBackend(query: string, method: string = "get", postData: string = null, errorHandler: any = null, errorHandlerParam: any = null) : Observable<any>
  {
    var _baseUrl = this.getBackendUrl();
    var _url = _baseUrl + query;

    console.log("Query: " + query);

    var _headers = new HttpHeaders(
    {
        'Content-Type': 'application/json'
    });

    if (this.token)
    {
      _headers = _headers.append("Authorization", "Bearer " + this.token);
    }

    var _options: any = 
    {
        headers: _headers,
        observe: 'body' as const,
        responseType: 'json' as const
    }      

    if (method == "get")
    {
      return this.http.get(_url, _options).pipe(catchError(errorHandler ? (error) => { return errorHandler(error, errorHandlerParam) } : this.handleError), finalize(() => {}));
    }
    else if (method == "post")
    {
      return this.http.post(_url, postData, _options).pipe(catchError(errorHandler ? (error) => { return errorHandler(error, errorHandlerParam) } : this.handleError));
    }
    else if (method == "put")
    {
      return this.http.put(_url, postData, _options).pipe(catchError(errorHandler ? (error) => { return errorHandler(error, errorHandlerParam) } : this.handleError));
    }    
    else if (method == "delete")
    {
      if (postData)
      {
        _options.body = postData;
        return this.http.request("delete", _url, _options).pipe(catchError(errorHandler ? (error) => { return errorHandler(error, errorHandlerParam) } : this.handleError));
      }
      else
      {
        return this.http.delete(_url, _options).pipe(catchError(errorHandler ? (error) => { return errorHandler(error, errorHandlerParam) } : this.handleError));
      }
    }
  }

  private getBackendUrl()
  {
    var androidLocalhost = "http://10.0.2.2:5000/";
    var iosLocalhost = "http://0.0.0.0:5000/";

    var url = androidLocalhost;

    if (isIOS)
    {
      url = iosLocalhost;
    }
    else if (isAndroid)
    {

    }

    return url;
  }

  private handleError(error: HttpErrorResponse) 
  {
    if (error.error && error.error.message) {
      // A client-side or network error occurred. Handle it accordingly.
      var msg = error.error.message;
      console.error(`An error occurred: ${msg}`);
      return throwError(msg);
    } else {
      // The backend returned an unsuccessful response code.
      // The response body may contain clues as to what went wrong.
      var baseUrl = error.url.split("?")[0];
      console.error(
        `Backend returned code ${error.status},\n` +
        `body was: ${error.error},\n` +
        `and URL was: ${baseUrl}`);
    }

    // Return an observable with a user-facing error message.
    return throwError('Error in backend.service, see details in console log.');
  }  
  
}

File: packages/fluttter/bookstore/lib/src/services/backend_service.dart
import 'dart:io';
import 'dart:convert';
import 'package:bookstore/src/data/library.dart';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'http_client_default.dart' if (dart.library.html) 'http_client_browser.dart';

class BackendService {

  String _errorMessage = "";
  String _token = "";
  String _cookie = "";
  String _serverUrl = "";

  String getBackendUrl() {
    if (_serverUrl.isEmpty) {

      String? urlFromSettings = Library.localStorage!.getString("ServerUrl");
      if (urlFromSettings != null)
      {
        return urlFromSettings;
      }

      String localhost = "http://localhost:5003/api/";
      String androidLocalhost = "http://10.0.2.2:5003/api/";
      String iosLocalhost = "http://0.0.0.0:5003/api/";

      if (kIsWeb)
      {
        _serverUrl = localhost;
      }
      else if (Platform.isAndroid) {
        _serverUrl = androidLocalhost;
      } else if (Platform.isIOS) {
        _serverUrl = iosLocalhost;
      } else {
        _serverUrl = localhost;
      }
    }

    return _serverUrl;
  }

  void updateBackendUrl(String url)
  {
    _serverUrl = url;
    Library.localStorage!.setString("ServerUrl", _serverUrl);
  }

  String getLastError()
  {
    return _errorMessage;    
  }

  void setLastError(String error)
  {
    _errorMessage = error;
  }

  void setToken(newToken)
  {
    _token = newToken;
  }

  void clearToken()
  {
    _token = "";
    _cookie = "";
  }

  Future<dynamic> _processResponse(http.Response response)
  {
    if (response.statusCode == 200) {

      if (response.headers.keys.contains("set-cookie"))
      {
        _cookie = response.headers["set-cookie"]!;
        print("Set-Cookie: $_cookie");

        int index = _cookie.indexOf(';');
        _cookie = (index == -1) ? _cookie : _cookie.substring(0, index);
      }

      if (response.body != "")
      {
        var jsonData = jsonDecode(response.body);
        return Future.value(jsonData);
      }
      return Future.value(0);
    } else {
      _errorMessage = response.reasonPhrase ?? "Response status code is ${response.statusCode}";
      print("Backend error: $_errorMessage");
      return Future.error(_errorMessage);
    }
  }

  void _processError(error)
  {
    if (error.message != null)
    {
      _errorMessage = error.message;
    }
    else
    {
      _errorMessage = error.toString();
    }
    print("Backend error: $_errorMessage");
    throw("Backend error");
  }

  Future<dynamic> getDataFromBackend(String query, [String method = "get", String? postData, Map<String, String>? additionalHeaders]) async {
    _errorMessage = "";
    var backendUrl = getBackendUrl();
    var url = backendUrl + query;

    print('Query: $query');

    Map<String, String> headers = {'Content-Type': 'application/json'};

    if (_token.isNotEmpty) {
      headers['Authorization'] = 'Bearer $_token';
    }

    if (_cookie.isNotEmpty) {
      headers['Cookie'] = _cookie;
    }

    if (additionalHeaders != null)
    {
       additionalHeaders.forEach((key, value) { 
         headers[key] = value;
       });

    }   
    
    var uri = Uri.parse(url);

    if (method == "get")
    {
      return client.get(uri, headers: headers).then(_processResponse).catchError(_processError);
    }
    else if (method == "delete")
    {
      return client.delete(uri, headers: headers).then(_processResponse).catchError(_processError);
    }
    else if (method == "put")
    {
      return client.put(uri, headers: headers, body: postData).then(_processResponse).catchError(_processError);
    }
    else if (method == "post")
    {
      return client.post(uri, headers: headers, body: postData).then(_processResponse).catchError(_processError);
    }    

  }
}
